src/common/Cache.ts

```diff
@@ -1,67 +1,152 @@
 import { Suggestion } from '@apis/CorrectionApi';
+import { TmdbApiConfig } from '@apis/TmdbApi';
+import { TraktSettingsResponse } from '@apis/TraktSettings';
+import { TraktHistoryItem } from '@apis/TraktSync';
+import { BrowserStorage } from '@common/BrowserStorage';
+import { SavedTraktItem } from '@models/TraktItem';
 
-export interface CacheValues {
-	suggestions: Partial<Record<string, Suggestion[] | null>>;
-	imageUrls: Partial<Record<string, string | null>>;
+export type CacheItems<T extends (keyof CacheValues)[]> = {
+	[K in T[number]]: CacheItem<K>;
+};
+
+export type CacheValues = {
+	[K in keyof CacheSubValues]: Partial<Record<string, Cacheable<CacheSubValues[K]>>>;
+};
+
+export interface CacheSubValues {
+	imageUrls: string | null;
+	itemsToTraktItems: string;
+	suggestions: Suggestion[] | null;
+	tmdbApiConfigs: TmdbApiConfig | null;
+	traktHistoryItems: TraktHistoryItem[];
+	traktItems: SavedTraktItem;
+	traktSettings: TraktSettingsResponse;
+	urlsToTraktItems: string;
 }
 
-class _Cache {
-	private timers: Record<keyof CacheValues, number | null> = {
-		suggestions: null,
-		imageUrls: null,
-	};
+export type Cacheable<T = unknown> = {
+	value: T;
+	timestamp: number;
+};
 
-	/** In seconds. */
-	private expiries: Record<keyof CacheValues, number> = {
-		suggestions: 360,
-		imageUrls: 360,
-	};
+export type CacheStorageValues = {
+	[K in `${keyof CacheValues}Cache`]?: CacheValues[CacheStorageKeysToKeys[K]];
+};
 
-	readonly values: CacheValues;
+export type CacheStorageKeysToKeys = ReverseMap<CacheKeysToStorageKeys>;
 
-	constructor() {
-		this.values = this.getInitialValues();
-		this.startTimers();
-	}
+export type CacheKeysToStorageKeys = {
+	[K in keyof CacheValues]: `${K}Cache`;
+};
+
+export class CacheItem<K extends keyof CacheValues> {
+	readonly cache: CacheValues[K];
 
-	private getInitialValues(): CacheValues {
-		return {
-			suggestions: {},
-			imageUrls: {},
-		};
+	constructor(cache: CacheValues[K]) {
+		this.cache = cache;
 	}
 
-	private invalidate<K extends keyof CacheValues>(key: K, expiry: number): void {
-		this.values[key] = this.getInitialValues()[key];
-		this.timers[key] = window.setTimeout(() => this.invalidate(key, expiry), expiry * 1e3);
+	get(subKey: string): CacheSubValues[K] | undefined {
+		return this.cache[subKey]?.value as never;
```

Comment:
The `never`?

Reply:
Yes

Reply:
I just couldn't figure out how to type it, because of the generics, so I went with `never` just to disable the error. I guess I could also do `as unknown as CacheSubValues[K]`. Is that better?

Reply:
I would assume never means it never reaches this code?

Reply:
I have another PR for better typings, so I'll fix this there.